### PR TITLE
Luya-ChartDashboardObject

### PR DESCRIPTION
**Pull request by baqianxin**
Usage:
```php
    public $dashboardObjects = [
        [
            'class' => 'luya\admin\dashboard\BasicDashboardObject',
            'outerTemplate' => '
                    <div class="card-image">
                        <img src="http://www.loveo.cc/wp-content/uploads/2017/02/card-image.jpg"
                             alt="Orange"/>
                        <div class="card-header"><h3>{{title}}</h3></div>
                    </div>
                    <div class=""><ul class="list-group list-group-flush">{{template}}</ul></div>',
            'template' => '<li class="list-group-item" ng-repeat="item in data"><p>{{item.salutation}}</p></li>',
            'dataApiUrl' => 'admin/api-addressbook-contact',
            'title' => 'The Users',
        ],
        [
            'class' => 'luya\admin\dashboard\ChartDashboardObject',
            'dataApiUrl' => 'admin/api-addressbook-contact/sale',
            'title' => 'The Chart1',
        ],
        [
            'class' => 'luya\admin\dashboard\ChartDashboardObject',
            'dataApiUrl' => 'admin/api-addressbook-contact/sale-by-wechat',
            'title' => 'The Chart2',
        ]
    ];
```
> The Echarts.js demo :+1: http://echarts.baidu.com/demo.html#scatter-polar-punchCard

The dataApi return：
```php
 return $option='{THE_ECHART_OPTION}';
```
example:
image
```php
//dataApi
  /**
     * @action sale 返回图表信息
     */
    public function actionSale(){
      $options = ' {
              "title": {
                  "text": "商城渠道"
              },
              "tooltip": {},
              "legend": {
                  "data":["销量"]
              },
              "xAxis": {
                  "data": ["衬衫","羊毛衫","雪纺衫","裤子","高跟鞋","袜子"]
              },
              "yAxis": {},
              "series": [{
                  "name": "销量",
                  "type": "bar",
                  "data": [5, 20, 36, 10, 10, 20]
              }]
          }';

      return $options;
    }
```

### CODE REVIEW

File: src/resources/scss/components/_card.scss [hunk 1]
```diff
@@ -27,4 +27,90 @@
...
+    transition-property: filter width;
+    transition-duration: .3s;
+}
+
+.card-body {
+    /*text-align: center;*/
+    padding: 15px 20px;
+    box-sizing: border-box;
+}
+
```
Comment (dev7ch) [inline src/resources/scss/components/_card.scss]: I am pretty unsure about the following generic card styles and fonts in general.

File: src/resources/scss/components/_card.scss [hunk 2]
```diff
@@ -27,4 +27,90 @@
...
+.card-title {
+    text-align: center;
+    text-transform: uppercase;
+    font-weight: bold;
+    letter-spacing: 1px;
+}
+
+.card-date,
+.card-exceprt,
+#case-title {
```
Comment (dev7ch) [inline src/resources/scss/components/_card.scss]: better do not use `#id` always try to use `class`

File: src/resources/scss/components/_card.scss [hunk 3]
```diff
@@ -27,4 +27,90 @@
...
     border-top: 1px solid $list-group-border-color;
-}
\ No newline at end of file
+}
+
+.card {
+    width: 100%;
+    margin: 10px auto;
+    background-color: white;
+    box-shadow: 0px 5px 20px #555;
```
Comment (dev7ch) [inline src/resources/scss/components/_card.scss]: @baqianxin this box-shadow will give style conflicts with the card´s in admin UI.
In general i would suggest to set the box-shadow not  as static css value, better a sass var like `$card-box-shadow`.

File: src/resources/js/directives.js [hunk 4]
```diff
@@ -1971,7 +1971,7 @@
             },
             template: function() {
                 return  '<div class="form-group form-side-by-side" ng-class="{\'input--hide-label\': i18n}">' +
-                            '<div class="form-side form-side-label">' +
+                            '<div class="form-control no-border">' +
```
Comment (dev7ch) [inline src/resources/js/directives.js]: the bs4 class `border-0` would do the trick.
see: https://getbootstrap.com/docs/4.0/utilities/borders/

**Comment by baqianxin** [comment]
ok,i will amend these

File: src/resources/js/directives.js (hunk 4 above)
Comment (baqianxin) [reply src/resources/js/directives.js]: thanks

File: src/resources/scss/components/_card.scss (hunk 2 above)
Comment (baqianxin) [reply src/resources/scss/components/_card.scss]: i agree whit you

**Comment by baqianxin** [comment]
How about this:
image

File: src/assets/Main.php [hunk 5]
```diff
@@ -27,6 +27,7 @@ class Main extends \luya\web\Asset
      */
     public $js = [
         'dist/js/main.js',
+        'https://cdn.bootcss.com/echarts/3.8.5/echarts-en.min.js',
```
Comment (TheMaaarc) [inline src/assets/Main.php]: We don't use CDN resources in the admin because we don't want to expose all the LUYA admin URLs to the respective CDN. Just integrate the JS via libs composer/bower (https://libraries.io/bower/echarts).

File: src/views/layouts/_angulardirectives.php [hunk 6]
```diff
@@ -147,16 +145,24 @@
         <div class="imageupload-preview" ng-show="imageinfo != null">
             <div class="imageupload-preview-sizer"></div>
             <img ng-src="{{thumb.source}}" ng-show="imageinfo != null" class="imageupload-preview-image" />
-            <div class="imageupload-infos">
-                <div class="imageupload-size" ng-show="!imageLoading">{{ imageinfo.resolutionWidth }} x {{ imageinfo.resolutionHeight }}</div>
+
```
Comment (nadar) [inline src/views/layouts/_angulardirectives.php]: This changes has nothing do to with your PR, you should make PR only about a single topic. Please reset this file to master.

File: src/resources/scss/components/_card.scss [hunk 7]
```diff
@@ -27,4 +27,45 @@
...
     border-top: 1px solid $list-group-border-color;
-}
\ No newline at end of file
+}
+
+.card {
+    background-color: white;
+}
+
+.card .card-header {
```
Comment (TheMaaarc) [inline src/resources/scss/components/_card.scss]: Why don't you add this css to `.card-header` at the first line of this file?

File: src/views/template/dashboard/chart.php [hunk 8]
```diff
@@ -0,0 +1,8 @@
+<?php
```
Comment (nadar) [inline src/views/template/dashboard/chart.php]: What is this empty file good for?

File: src/resources/scss/components/_zaa-image-array-upload.scss [hunk 9]
```diff
@@ -12,4 +12,28 @@ zaa-image-array-upload {
     & + .flag {
         margin-right: 0;
     }
-}
\ No newline at end of file
+}
+
+.no-border{
```
Comment (nadar) [inline src/resources/scss/components/_zaa-image-array-upload.scss]: This changes has nothing do to with your PR, you should make PR only about a single topic. Please reset this file to master.

File: src/resources/scss/components/_card.scss [hunk 10]
```diff
@@ -27,4 +27,45 @@
 
 .card > .card-body + .list-group > .list-group-item:first-of-type {
     border-top: 1px solid $list-group-border-color;
-}
\ No newline at end of file
+}
```
Comment (nadar) [inline src/resources/scss/components/_card.scss]: This changes has nothing do to with your PR, you should make PR only about a single topic. Please reset this file to master.

File: src/resources/scss/components/_card.scss [hunk 11]
```diff
@@ -27,4 +27,45 @@
...
+
+.card {
+    background-color: white;
+}
+
+.card .card-header {
+    position: relative;
+    padding: .75rem 1.25rem;
```
Comment (TheMaaarc) [inline src/resources/scss/components/_card.scss]: As I stated in the other PR - Use px for padding - our Gulp workflow automatically converts them to rem and we still have px as fallback for browsers that don't support rem.

File: src/resources/js/directives.js [hunk 12]
```diff
@@ -2,7 +2,50 @@
...
+            item: "=",
+            data: "="
+        },
+        restrict: 'E',
+        template: '<div style="min-height:300px;height:auto;width:100%;"></div>',
+        replace: true,
+        link: function($scope, element, attrs, controller) {
+            var myChart = echarts.init(document.getElementById($scope.id),'macarons');
+            var option = {
+                // 提示框，鼠标悬浮交互时的信息提示
```
Comment (nadar) [inline src/resources/js/directives.js]: please use english comments using /* */ instead of // due to minification

File: src/resources/scss/components/_card.scss [hunk 13]
```diff
@@ -27,4 +27,45 @@
...
+    margin-bottom: 0;
+    background-color: rgba(0,0,0,0.03);
+    border-bottom: 1px solid rgba(0,0,0,0.125);
+}
+.card .card-header h3{
+    color: #fff;
```
Comment (TheMaaarc) [inline src/resources/scss/components/_card.scss]: Again, as I said in the other PR - Use the color function to get a specific color (that way we don't have XXX shades of colors in the admin). See: https://getbootstrap.com/docs/4.0/getting-started/theming/#functions

File: src/resources/scss/components/_card.scss [hunk 14]
```diff
@@ -27,4 +27,45 @@
...
+    font-weight: 400;
+}
+.card a {
+    color: #333;
```
Comment (TheMaaarc) [inline src/resources/scss/components/_card.scss]: `color:` https://getbootstrap.com/docs/4.0/getting-started/theming/#functions

File: src/resources/js/directives.js [hunk 15]
```diff
@@ -1971,7 +2014,7 @@
             },
             template: function() {
                 return  '<div class="form-group form-side-by-side" ng-class="{\'input--hide-label\': i18n}">' +
-                            '<div class="form-side form-side-label">' +
+                            '<div class="form-control border-0">' +
```
Comment (nadar) [inline src/resources/js/directives.js]: This should not be part of this PR.

File: .gitignore [hunk 16]
```diff
@@ -1,5 +1,7 @@
+/scripts
```
Comment (nadar) [inline .gitignore]: Why adding scripts to ignore list?

File: src/resources/scss/components/_card.scss [hunk 17]
```diff
@@ -27,4 +27,45 @@
...
+    text-decoration: none;
+}
+
+.card:hover .card-image img {
+    width: 110%;
```
Comment (TheMaaarc) [inline src/resources/scss/components/_card.scss]: Why `width: 110%;`?

File: src/resources/scss/components/_zaa-image-array-upload.scss [hunk 18]
```diff
@@ -12,4 +12,28 @@ zaa-image-array-upload {
...
+.imageupload-preview{
+  margin-bottom: 10px;
+}
+.imageupload-preview-image{
+  top: 46%;
+}
+.imageupload-size {
+    background-color:transparent;// #E50060;
+    bottom:auto;// -30px;
+    color:#000;// #fff;
```
Comment (TheMaaarc) [inline src/resources/scss/components/_zaa-image-array-upload.scss]: color: https://getbootstrap.com/docs/4.0/getting-started/theming/#functions

File: src/views/layouts/_angulardirectives.php [hunk 19]
```diff
@@ -126,9 +126,7 @@
                     <?= Admin::t('layout_select_file'); ?>
                 </span>
             </div>
-            <span class="link-selector-reset" ng-click="reset()" ng-show="fileinfo!=null">
-                <i class="material-icons">remove_circle</i>
-            </span>
+            <button type="button" class="btn btn-danger" ng-click="reset()" ng-show="fileinfo!=null" ><?= Admin::t('aws_groupauth_th_remove');?></button>
```
Comment (TheMaaarc) [inline src/views/layouts/_angulardirectives.php]: As I said in the other PR - Use `btn btn-delete btn-icon` instead of `btn btn-danger`

File: .gitignore (hunk 16 above)
Comment (baqianxin) [reply .gitignore]: i want rebase the master,and remove the extra commits

File: src/views/template/dashboard/chart.php (hunk 8 above)
Comment (baqianxin) [reply src/views/template/dashboard/chart.php]: Please wait, I will commit again

File: src/resources/scss/components/_card.scss (hunk 17 above)
Comment (baqianxin) [reply src/resources/scss/components/_card.scss]: Mouse over will have a magnifying effect

File: src/views/layouts/_angulardirectives.php (hunk 19 above)
Comment (baqianxin) [reply src/views/layouts/_angulardirectives.php]: i  will try again as you said，thanks

File: .gitignore (hunk 16 above)
Comment (baqianxin) [reply .gitignore]: image

File: src/resources/scss/components/_card.scss (hunk 17 above)
Comment (TheMaaarc) [reply src/resources/scss/components/_card.scss]: Ahhh - sorry, didn't see that `:hover`. 👍

File: src/views/layouts/_angulardirectives.php (hunk 19 above)
Comment (baqianxin) [reply src/views/layouts/_angulardirectives.php]: @TheMaaarc Although I also think I changed the look better

Comment (nadar) [reply src/views/layouts/_angulardirectives.php]: @baqianxin I like the shadows as well, it looks good. but we should be consistent according to bootstrap 4 - sometimes this is more important

Comment (nadar) [reply src/views/layouts/_angulardirectives.php]: @baqianxin For us consistency and the close styles to bootstrap 4 without changing base css is important for us. This way you will be able to apply a bootstrap4 theme to the admin you complet change the look and feel!
